### ISSUE TITLE
Make gradle-wrapper more robust

### ISSUE THREAD

**Issue by jonnyzzz** [gradle/gradle#6012]
I had several crashes with gradle wrapper. Most of them were in the CI environment, where machines are resued and I'm not able to control the state

```
Starting: "C:\Build Agent\work\7cd074a3920f1b82\gradlew.bat" --init-script "C:\Build Agent\plugins\gradle-runner\scripts\init.gradle" -Dorg.gradle.daemon=false -b build.gradle teamcityBuild
[10:48:56][Step 1/1] in directory: C:\Build Agent\work\7cd074a3920f1b82
[10:48:57][Step 1/1] Downloading https://cache-redirector.jetbrains.com/services.gradle.org/distributions/gradle-4.9-all.zip
[10:49:00][Step 1/1] ..............
[10:49:00][Step 1/1] Could not unzip C:\Users\builduser\.gradle\wrapper\dists\gradle-4.9-all\7ij6mqkzis5urpjf22tuuh315\gradle-4.9-all.zip to C:\Users\builduser\.gradle\wrapper\dists\gradle-4.9-all\7ij6mqkzis5urpjf22tuuh315.
[10:49:00][Step 1/1] Reason: error in opening zip file
[10:49:00][Step 1/1] Exception in thread "main" java.util.zip.ZipException: error in opening zip file
[10:49:00][Step 1/1] 	at java.util.zip.ZipFile.open(Native Method)
[10:49:00][Step 1/1] 	at java.util.zip.ZipFile.<init>(ZipFile.java:225)
[10:49:00][Step 1/1] 	at java.util.zip.ZipFile.<init>(ZipFile.java:155)
[10:49:00][Step 1/1] 	at java.util.zip.ZipFile.<init>(ZipFile.java:169)
[10:49:00][Step 1/1] 	at org.gradle.wrapper.Install.unzip(Install.java:218)
[10:49:00][Step 1/1] 	at org.gradle.wrapper.Install.access$600(Install.java:27)
[10:49:00][Step 1/1] 	at org.gradle.wrapper.Install$1.call(Install.java:75)
[10:49:00][Step 1/1] 	at org.gradle.wrapper.Install$1.call(Install.java:48)
[10:49:00][Step 1/1] 	at org.gradle.wrapper.ExclusiveFileAccessManager.access(ExclusiveFileAccessManager.java:69)
[10:49:00][Step 1/1] 	at org.gradle.wrapper.Install.createDist(Install.java:48)
[10:49:00][Step 1/1] 	at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:107)
[10:49:00][Step 1/1] 	at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:61)
[10:49:00][Step 1/1] Process exited with code 1
[10:49:01][Step 1/1] Process exited with code 1 (Step: Gradle)
```
For some reason, the file in caches is corrupted. It might be caches issues or wrapper process was simpy killed. It looks good for the gradle wrapper to re-try download if the previous one failed to open.

I decided to fix the problem by changing paths in `gradle-wrapper.properties`:
```properties
distributionBase=GRADLE_USER_HOME
distributionPath=wrapper/dists-vcs
distributionUrl=https\://cache-redirector.jetbrains.com/services.gradle.org/distributions/gradle-4.9-all.zip
zipStoreBase=GRADLE_USER_HOME
zipStorePath=wrapper/dists-vcs
```

And it started to fail as 
```
Gradle distribution '/Users/jonnyzzz/.gradle/wrapper/dists-vcs/gradle-4.9-all/7ij6mqkzis5urpjf22tuuh315' does not contain any directories. Expected to find exactly 1 directory.
```
Both failures make it harder to use Gradle on CI or developer machines. It might be good to add more stability and resiliency to sudden failures of Gradle wrapper.

**Comment by marcphilipp** [comment]
If I understand you correctly, you're suggesting that the wrapper should redownload the ZIP it just downloaded? Why would it do that? It expects a distribution to be versioned and, as such, unmodifiable. You may also [configure the wrapper to verify the checksum](https://docs.gradle.org/current/userguide/gradle_wrapper.html#sec:verification) of the downloaded ZIP.

**Comment by jonnyzzz** [comment]
I will not agree with you. You are not able to control the machine as a whole thing. Your process may be killed spontaneously. Assertions are good, but it creates layers of a hassle to others who got stuck with them. For example, the problems I reported I saw on CI machines. We have a pool of such machines, hundreds of them, and some now stopped to run Gradle. For many projects. It is HARD to ssh to each of them to have that trivial problem fixed. BTW, such problems typically ruin the first user experience.

Technically, what may go wrong
 - a machine may crash suddenly
 - Gradle and Gradle Wrapper process may be killed suddenly
 - a disk may be cleaned up suddenly or corrupted
 - network error may end up a broken file is downloaded

And yes, the broken Gradle file is what I had in my case. Again, with CI machines. Again it is hard to detect those machines and clean those folders. 

It does not feel like a rocket science to add some trivial assertions WITH means to self-healing. It will just make our lives easier and without dumd errors which are time-consuming to fix on and on again.
